Objective-C/CBLCoreMLPredictiveModel.m

```diff
@@ -0,0 +1,538 @@
+                // Make sure output type is image
+                if (desc.outputDescriptionsByName.allValues[0].type == MLFeatureTypeImage) {
+                    CVPixelBufferRef pixel = ((VNPixelBufferObservation*)o).pixelBuffer;
+                    value = [self.class blobFromPixelBuffer: pixel];
+                }
+            }
+        }
+        
+        if (!value)
+            return nil;
+        
+        CBLMutableDictionary* result = [[CBLMutableDictionary alloc] init];
+        [result setValue: value forKey: outputKeys[0]];
+        return result;
+    }
+}
+
+
+#pragma mark - CBLDictionary to Features Conversion
+
+
++ (id<MLFeatureProvider>) featuresFromDictionary: (CBLDictionary*)dictionary
+                                            spec: (NSDictionary<NSString*, MLFeatureDescription*>*)spec {
+    NSMutableDictionary* dict = [[NSMutableDictionary alloc] initWithCapacity: spec.count];
+    for (NSString* key in spec) {
+        MLFeatureDescription* desc = spec[key];
+        id value = [dictionary valueForKey: key];
+        if (!value) {
+            if (desc.optional)
+                continue;
+            CBLWarn(Query, @"The value of the non-optional input key (%@) is missing.", key);
+            return nil;
+        }
+        
+        BOOL isSequenceValue = NO;
+        MLFeatureType sequenceType = MLFeatureTypeInvalid;
+        if (@available(macOS 10.14, iOS 12.0, *)) {
+            isSequenceValue = (desc.type == MLFeatureTypeSequence);
+            sequenceType = desc.sequenceConstraint.valueDescription.type;
+        }
+        
+        MLFeatureValue* featureValue;
+        if (isSequenceValue) {
+            if (@available(macOS 10.14, iOS 12.0, *)) {
+                featureValue = [self.class sequenceFeatureValueFromValue: value type: sequenceType];
+            }
+        } else if (desc.type == MLFeatureTypeMultiArray) {
+            MLMultiArrayConstraint* constraint = desc.multiArrayConstraint;
+            featureValue = [self.class multiArrayFeatureValueFromValue: value
+                                                                 shape: constraint.shape
+                                                                  type: constraint.dataType] ;
+        } else {
+            featureValue = [self.class featureValueFromValue: value type: desc.type];
+        }
+        
+        if (!featureValue) {
+            CBLWarn(Query, @"Cannot convert the value of %@ key.", key);
+            return nil;
+        }
+        [dict setObject: featureValue forKey: key];
+    }
+    
+    NSError* error;
+    MLDictionaryFeatureProvider* provider =
+        [[MLDictionaryFeatureProvider alloc] initWithDictionary: dict error: &error];
+    if (!provider) {
+        CBLWarn(Query, @"Error when creating a dictionary feature provider: %@", error);
+    }
+    return provider;
+}
+
+
++ (MLFeatureValue*) featureValueFromValue: (id)value type: (MLFeatureType)type {
```

Comment:
return type seems like nullable, wonder why it is not shown on Xcode compiler.

```suggestion
+ (nullable MLFeatureValue*) featureValueFromValue: (id)value type: (MLFeatureType)type {
```

Reply:
We have it defined as nullable in .h.